File: sort/shell/shell.cpp
#include <vector>
#include <iostream>
#include <math.h>

using namespace std;

void shellSort(vector<int>*);

void shellSort(vector<int>* arr) {
    vector<int>& A = *arr;
    // Calculate the gap initially, this can be arr / 2 or 3
    int gap = static_cast<int>(floor(arr->size() / 2));
    // We want to calculate the gap to 1
    while (gap >= 1) {
        // Count up so we have a sliding window
        for (int i = 0; i < static_cast<int>(arr->size()) - gap; i++) {
            int pos = i + gap;
            // For each sliding window position filter the lowest value down the chain
            while (pos-gap >= 0 && A[pos] < A[pos - gap]) {
                swap(A[pos], A[pos-gap]);
                pos -= gap;
            }
        }
        gap = gap / 2;
    }
}

//vector<int> nums { 8, 9, 7, 5, 2, 4, 6, 1, 3, -1 };
//shellSort(&nums);
//printVector(nums);

File: sort/shell/shell.java
public class Main {

    static void shellSort(int[] arr) {

        int gap = (int)Math.floor(arr.length / 2);
        while (gap >= 1) {
            for (int i = 0; i < arr.length - gap; i++) {
                int pos = i + gap;
                while (pos - gap >= 0 && arr[pos] < arr[pos - gap]) {
                    // Java has no swap syntax for arrays, only lists
                    int temp = arr[pos];
                    arr[pos] = arr[pos - gap];
                    arr[pos - gap] = temp;
                    pos -= gap;
                }
            }
            gap = (int)Math.floor(gap / 2);
        }
    }

    public static void main(String[] args) {
        int[] arr =  {8, 9, 7, 5, 2, 4, 6, 1, 3, -1};
        shellSort(arr);
        for(int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        assert arr == new int[] {-1, 1, 2, 3, 4, 5, 6, 7, 8, 9};
    }
}
